### ISSUE TITLE
Sound support

### ISSUE THREAD

**Issue by CedricGuillemet** [BabylonJS/BabylonNative#31]
crossplatform audio engine
https://github.com/jarikomppa/soloud
Do you know any other one?

**Comment by bghgary** [comment]
https://github.com/resonance-audio/resonance-audio

**Comment by CedricGuillemet** [comment]
Some more tiny libraries here
https://github.com/nothings/single_file_libs#audio

**Comment by bghgary** [comment]
From forum post: https://forum.babylonjs.com/t/babylonnative-questions/9716

https://github.com/LabSound/LabSound

**Comment by chrisfromwork** [comment]
some real time filter libraries: 

https://github.com/vinniefalco/DSPFilters
https://github.com/berndporr/iir1

**Comment by chrisfromwork** [comment]
https://github.com/andrewrk/libsoundio

**Comment by chrisfromwork** [comment]
https://github.com/cycfi/q

**Comment by Drigax** [comment]
We may also want to consider:
https://www.videolan.org/vlc/libvlc.html

**Comment by CedricGuillemet** [comment]
@Drigax Looks interesting for video playback support as well. I'm curious about the footprint on mobile.

**Comment by Drigax** [comment]
Project   Name | Link To Project | Does it implement | What does it do? | Oses Supported: | License | Notes
-- | -- | -- | -- | -- | -- | --
  |   | Audio Backend | DSP | Modular effect chaining |   | Windows/Win32 | Windows/UWP | Android | MacOS | iOS | Linux |   |  
c | https://source.chromium.org/chromium/chromium/src | ✅ | ✅ | ✅ | Full browser implementation.   Includes implementations for WebAudio API with platform audio backends | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ |   | This is our gold standard   (because it pretty much is the standard.) Supports all the funtionality of   WebAudio without additional dependencies, and works as a full multimedia   backend.
SoLoud | https://github.com/jarikomppa/soloud | ❌ | ✅ | ❌ | Portable audio processing library   with extensible backend support | ❔ | ❔ | ❔ | ❔ | ❔ | ❔ | Zlib | This isn't necessarily an   all-in-one solution, as it relies on pulling in an audio backend as a   dependency. However it has a number of audio effects implemented to support   the subset of WebAudio that we need to implement.
libSDL | https://github.com/libsdl-org/SDL | ✅ | ❌ | ❌ | Low level cross-platform I/O   interface | ✅ | ❔ | ✅ | ✅ | ✅ | ✅ | Zlib | This is more than just a cross   platform audio backend (and interface to other desired backends), this   library aims to solve multiple cross platform I/O issues such as   input/haptics, graphics, sensors, file system access, etc…
Resonance Audio | https://github.com/resonance-audio/resonance-audio | ❌ | ✅ | ✅ | Audio processing library with   node graph and a wide library of built-in DSP effects | ❔ | ❔ | ❔ | ❔ | ❔ | ❔ | Apache 2 | This is a DSP middleware   package. While it should support the functionality of WebAudio, it requires   pulling in an audio backend implementation
JACK | https://github.com/jackaudio/jack2 | ❌ | ❌ | ❌ | Audio server |   |   |   |   |   |   |   |  
  |   |   |   |   |   |   |   |   |   |   |   |   |  
  |   |   |   |   |   |   |   |   |   |   |   |   |  
Key: |   |   |   |   |   |   |   |   |   |   |   |   |  
❔ | Not conclusive, need more   information |   |   |   |   |   |   |   |   |   |  
✅ | supported /included |   |   |   |   |   |   |   |   |   |   |  
❌ | Not supported / not   included |   |   |   |   |   |   |   |   |   |   |

**Comment by Drigax** [comment]
I've spent the past few days researching various popular audio solutions, and have tabulated some of their qualities  into an easy to digest chart. This should allow us to compare solutions/a set of solutions to see what direction we want to go with regarding incorporating existing audio projects.

**Comment by Drigax** [comment]
Currently we are still investigating the feasibility of incorporating Chromium build artifacts that contain WebAudio, and the Chrome media system libraries and dependencies. Ashu Tatake has taken the first steps in this approach.

@ryantrem had a very interesting concern regarding this first option, Apple has been known to inspect and remove apps that contain Chromium source. We should look into which apps were known to be removed for this reason, if there was any justification provided, and whether Apple has a clear policy on whether it is a blanket ban on any chromium-derived binaries, or specific parts of Chromium which are not allowed.

I'll follow up by investigating this as well.

**Comment by Drigax** [comment]
I've written a blog post here, outlining the plan moving forward for this feature:
https://babylonjs.medium.com/audio-in-babylon-native-charting-the-way-c6bc6d5c0d74

**Comment by Drigax** [comment]
Update 2021-05-17

The project and samples were successfully built on Windows, it may be worth taking the next step and getting the project to build  as a part of BabylonNative's build. 

I'll have to look into how to isolate building the engine+dependencies as a consumable lib without building the additional samples

**Comment by Drigax** [comment]
Update 2021-06-07

My work is currently staged on: 
https://github.com/Drigax/BabylonNative/tree/add-labsound (BabylonNative)
https://github.com/Drigax/Babylon.js/tree/add-native-audio (Babylon.js)

The native changes currently implement a Plugin component for NativeAudio, and modifies the Win32 playground to initialize NativeAudio. Currently the NativeAudio plugin intializes N-Api wrappers for the WebAudio classes needed to support Babylon.js's Sound and AudioEngine classes.

In Babylon.js some changes were made to engine instantiation so that the NativeEngine can have its AudioEngine initialized using the Engine constructor, also the AudioEngine was copied and modified to be usable by the NativeEngine (removing all dom-related functionality) as the NativeAudioEngine implementation.

**Comment by Drigax** [comment]
Moving forward there are some outstanding work to be done until I believe this work is ready for merging:

- [ ] Remove debug logging from NativeAudio.cpp
      - [ ] Ideally, refactor to use a preprocessor macro/AppRuntime level logging component.
- [ ] Move NativeAudio.cpp class/member declarations to a header, and refactor method implementations in respect of these changes.
- [ ] Implement _some_ code reuse regarding AudioNode derived object wrappers. Currently, N-API doesn't implement javascript inheritance for ObjectWrappers. Also, directly inheriting from the ObjectWrapper-derived base class in C++ seems to be a dead end. I was able to work around this through explicitly implementing the `AudioNode` contract in `GainNode` (with the intention of using this for the other node implementations, however this may cause some unexpected side effects as we implement more of the contract....
- [ ] Begin mapping N-API object wrappers to Labsound objects. So far it seems that we can 1:1 map an object wrapper to LabSounds equivalent, initializing the Labsound object on objectwrapper constructor.
- [ ] Finish stubbing the rest of the WebAudio contract. While not needed to fulfill this task, it would lay the groundwork for expanding future audio functionality easily (and providing functionality outside of the Babylon.js contract)

**Comment by bghgary** [comment]
As noted here (https://forum.babylonjs.com/t/source-for-n-api-implemented-bindings-of-web-audio-found/22132), might be interesting to check out this also:

https://github.com/node-3d/webaudio-raub

**Comment by FranticOwl** [comment]
@bghgary, more than a year later and this is still an open issue?

Is there any actual plan to build support for Babylon Native, so that frameworks such as React-Native can simply use the sound API from Babylon to add audio to scenes?

**Comment by bghgary** [comment]
Unfortunately, sound support is not actively being worked on right now. We attempted to start this a couple of times, but they both didn't work out for various reasons. cc @docEdub

**Comment by rootVIII** [comment]
Would love to see audio functionality added to this project.. even if only for a few select targets/OSs.

**Comment by markgrossnickle** [comment]
No ability to play audio seems like a deal breaker for many game projects. It's why were are going to go with another platform fwiw.

**Comment by matthargett** [comment]
As we approach the next phase of our product, which includes bringing Babylon React Native to visionOS, Spatial Audio is a need that we have. What can we do at Rebecker Specialties to help close this key gap that also affects using Babylon React Native on AOSP-based headsets (Pico 4, HTC Vive XR/Focus, etc)?

**Comment by bghgary** [comment]
Let's check with @docEdub. He was looking at this at some point. Any thoughts @docEdub?

**Comment by CedricGuillemet** [comment]
Valve Steam audio supports major platforms and is apache 2.0
https://github.com/ValveSoftware/steam-audio

**Comment by docEdub** [comment]
Yes, thanks for your interest! The new audio engine for Bablyon.js has been released and one of its design requirements is that it be easy to implement in BabylonNative. This is a big step forward!

That being said, I don't have the bandwidth to implement BabylonNative audio, yet, so help from the community would be much appreciated!

For anybody else reading this, comments and a +1 on the issue description help us determine priorities so smash that thumbs-up button 👍 on the first comment 👍
